ChatModel: Greatly improve parsing

kristofbolyai opened the pull request:
Fixes #334. 

This PR implements lossless processing for chat messages, fixing the issue where clickables are not working (mainly fixing dialogues where the user needs to pick an option).

This PR also fixes the issue where non-progressable quest dialogue was not displayed on the NPC dialogue overlay. 

This does not implement #292, but fixes issues related to it in ChatModel.

Incompleteusern commented:
Perhaps Add totem to the Presets too

magicus commented:
Thanks!

kristofbolyai committed:
ChatModel: Encode components losslessly in multi-line chat messages

kristofbolyai committed:
ChatModel: Fix not auto-completable dialogues not tracking

magicus commented:
Ooooh! Exciting. :-) I'm looking forward to going through this in detail. I'm also very happy that you are helping out with the chat stuff, it's better to have shared knowledge of all parts of the code and I started feeling a bit lonely in the chat corner. :)

kristofbolyai commented:
> Ooooh! Exciting. :-) I'm looking forward to going through this in detail. I'm also very happy that you are helping out with the chat stuff, it's better to have shared knowledge of all parts of the code and I started feeling a bit lonely in the chat corner. :)

I edited the description. There is no lossless encoding of components to strings, but I did fix issues lossless encoding could have fixed in ChatModel. Still, I think this is a pretty good fix.

kristofbolyai committed:
Merge branch 'main' into clickable_dialogue_fix

magicus commented:
> fixing dialogues where the user needs to pick an option

Did you find any such examples in the wild? I've not seen any since Winter Detlas. (I might have asked this before...)

kristofbolyai commented:
> > fixing dialogues where the user needs to pick an option
> 
> 
> 
> Did you find any such examples in the wild? I've not seen any since Winter Detlas. (I might have asked this before...)

https://wynncraft.fandom.com/wiki/The_Canary_Calls

Step 2 has both options and non-shiftable dialogue.

magicus commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
I'm seriously not really understanding how this can work. From what I remember, the "\n" was embedded in the middle of messages, and the "sibling" relationship segmented the message according to formatting instead. That was kind of the reason I started using formatted strings instead of Components. 

How do you ensure that `lines` contains just "lines"?

magicus commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
What do you think about `nonBlockingDialogue` as term instead?

kristofbolyai commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
From what I've seen, when the chat contents are resent, it is a huge, empty TextComponent, with each message being a sibling (these components might have additional siblings, but we do not care about them). As we do not split on "\n",  I needed to change a few places where we expected empty lines, etc. This seemed to work fine.

kristofbolyai committed:
ChatModel: Rename variable

magicus commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
There used to be an empty line inbetween. Like:
```
xxxxx: ....

[1/3] Key Master: Are you the gatekeeper?

Press SHIFT to continue
```

We needed to remove that, both to trim down the message sent as an event, and to avoid falsely triggering the detection of the "end" (that is, the start, since we're doing this in reverse) of the NPC message.

I don't understand how that is handled if you remove this?

kristofbolyai commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
We remove all empty lines before this code is reached.

magicus commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
Here, too, I think I prefer `nonBlocking`. (Or, if you want to, inverse the logic and use `blocking`)

magicus commented on common/src/main/java/com/wynntils/features/user/overlays/NpcDialogueOverlayFeature.java:
Shouldn't this test be inverted? If this means "NPC chat event is nonBlocking", then we should not schedule a sneak key press...

magicus commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
This might have been a change with 2.0 then. I think I'll need to verify this myself. It is indeed very welcome if this is the case. I am 99% certain it did not look like that when I created this functionality.

kristofbolyai commented on common/src/main/java/com/wynntils/features/user/overlays/NpcDialogueOverlayFeature.java:
Oops

magicus commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
```suggestion
                // Remove all empty lines
                if (noCodes.isBlank()) continue;

```

kristofbolyai committed:
Add comment

Co-authored-by: Magnus Ihse Bursie <mag@icus.se>

kristofbolyai committed:
NpcDialogueOverlayFeature: Fix variables

kristofbolyai commented:
I did notice a bug with dialogue parsing. Nothing serious, I think siblings will still work. I will try to fix it soon.

kristofbolyai committed:
ChatModel: Remove use cases of empty line parsing

kristofbolyai commented:
> I did notice a bug with dialogue parsing. Nothing serious, I think siblings will still work. I will try to fix it soon.

Fixed now. I also removed some empty line filtering logic, it is not needed anymore.

magicus commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
Well, it is still the case that `\n` are embedded. But perhaps it does not matter:

```
TextComponent{text='
                       ', siblings=[TextComponent{text='Welcome to Wynncraft!
', siblings=[], style=Style{ color=dark_gray, bold=true, italic=null, underlined=null, strikethrough=null, obfuscated=null, clickEvent=null, hoverEvent=null, insertion=null, font=minecraft:default}}, TextComponent{text='               ', siblings=[], style=Style{ color=gray, bold=null, italic=null, underlined=null, strikethrough=null, obfuscated=null, clickEvent=null, hoverEvent=null, insertion=null, font=minecraft:default}}, TextComponent{text='play.wynncraft.com', siblings=[], style=Style{ color=gray, bold=null, italic=null, underlined=null, strikethrough=null, obfuscated=null, clickEvent=ClickEvent{action=OPEN_URL, value='http://play.wynncraft.com'}, hoverEvent=null, insertion=null, font=minecraft:default}}, TextComponent{text=' ', siblings=[], style=Style{ color=gray, bold=null, italic=null, underlined=null, strikethrough=null, obfuscated=null, clickEvent=null, hoverEvent=null, insertion=null, font=minecraft:default}}, TextComponent{text='-/-', siblings=[], style=Style{ color=dark_gray, bold=null, italic=null, underlined=null, strikethrough=null, obfuscated=null, clickEvent=null, hoverEvent=null, insertion=null, font=minecraft:default}}, TextComponent{text=' wynncraft.com

', siblings=[], style=Style{ color=gray, bold=null, italic=null, underlined=null, strikethrough=null, obfuscated=null, clickEvent=null, hoverEvent=null, insertion=null, font=minecraft:default}}, TextComponent{text='                       ', siblings=[], style=Style{ color=gray, bold=null, italic=null, underlined=null, strikethrough=null, obfuscated=null, clickEvent=null, hoverEvent=null, insertion=null, font=minecraft:default}}, TextComponent{text='Wynncraft 2.0 is OUT!
', siblings=[], style=Style{ color=gray, bold=true, italic=null, underlined=null, strikethrough=null, obfuscated=null, clickEvent=null, hoverEvent=null, insertion=null, font=minecraft:default}}, TextComponent{text='                     ', siblings=[], style=Style{ color=gray, bold=null, italic=null, underlined=null, strikethrough=null, obfuscated=null, clickEvent=null, hoverEvent=null, insertion=null, font=minecraft:default}}, TextComponent{text='Read about the new abilities
', siblings=[], style=Style{ color=gray, bold=null, italic=null, underlined=null, strikethrough=null, obfuscated=null, clickEvent=null, hoverEvent=null, insertion=null, font=minecraft:default}}, TextComponent{text='                         ', siblings=[], style=Style{ color=gray, bold=null, italic=null, underlined=null, strikethrough=null, obfuscated=null, clickEvent=null, hoverEvent=null, insertion=null, font=minecraft:default}}, TextComponent{text='and archetypes ', siblings=[], style=Style{ color=gray, bold=null, italic=null, underlined=null, strikethrough=null, obfuscated=null, clickEvent=null, hoverEvent=null, insertion=null, font=minecraft:default}}, TextComponent{text='here!', siblings=[], style=Style{ color=gray, bold=null, italic=null, underlined=true, strikethrough=null, obfuscated=null, clickEvent=null, hoverEvent=null, insertion=null, font=minecraft:default}}, TextComponent{text='

', siblings=[], style=Style{ color=gray, bold=null, italic=null, underlined=null, strikethrough=null, obfuscated=null, clickEvent=null, hoverEvent=null, insertion=null, font=minecraft:default}}], style=Style{ color=null, bold=null, italic=null, underlined=null, strikethrough=null, obfuscated=null, clickEvent=null, hoverEvent=null, insertion=null, font=minecraft:default}}
```

Edit: I tried copy/paste from IntelliJ watch but it replaced the `\n` as shown in the watch window with real line breaks, so it just looks odd here. Every new line here is a "\n" in the text. And this is an entire sibling of the top-level TextComponent.

kristofbolyai commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
When the siblings have newlines, then it is some special Wynncraft message. I think it is fine to keep them as a "bunch" (for our purpose of quest dialogue parsing).

magicus commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
All "normal" chat history (user sent messages) seem to have one sibling per line. So these "multiline" advertising messages perhaps don't matter that they are several lines long.

But... I wonder what happens when they are starting to creep out of the history range. Can it lead us to fail to detect the "last message"?

I am seriously worried a bit that this entire change is making the chat manager less robust. If nothing else, we will now be *relying* on the assumption that there will be basically one sibling per line (or several-lines-long-messages). If this breaks for some reason, due to chat message types we have not yet spotted, the entire logic can just come crumbling down, and deliver very weird results to the user.

magicus commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
To feel 100% confident, I'd rather see that we transformed the TextComponent tree, so each top-level sibling contained exactly one line. For the normal chat messages, this is already the case, so we need not do anything there.

kristofbolyai commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
This change might make us more reliant on Wynn in a way yes. But, we used to rely on newlines being in specific places, in a specific format (always newline before "Press shift to continue", etc.). I think this state of ChatModel is almost as "breakable" as the old version.

kristofbolyai commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
> To feel 100% confident, I'd rather see that we transformed the TextComponent tree, so each top-level sibling contained exactly one line. For the normal chat messages, this is already the case, so we need not do anything there.

I am not sure doing this would be any better, as there is a really limited amount of messages which still have new lines, and even if they exist, they do not break our parsing system.

magicus commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
Some pseudo code on how I'm thinking:
```
TextComponent newComponent = new MutableComponent();
for sibling : TextComponent.getSiblings() {
  String text = sibling.getString();
  if (text.indexOf("\n") == text.length()) {
    // all is fine
   newComponent.add(sibling);
    continue
  }
  // else we need to either split this sibling up, or join it with the next, or possiblt both
 doTrickyStuff();
}
textComponent = newComponent;
```
The only really complicated thing is the `doTrickyStuff()`....

kristofbolyai commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
Also, technically, if we do this, this should be able to recurse on all siblings of all siblings. 

Still, I am not sure why we are trying to "optimize" the Component representation of a message (it feels like we are doing that). We could do this, but, either way, both the optimized and the normal component has the full information of a "message" (even if the said message is multi-line).

magicus commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
> I am not sure doing this would be any better, as there is a really limited amount of messages which still have new lines, and even if they exist, they do not break our parsing system.

Maybe. I'm not fully convinced. I'm basically worrying about two things:

1) That there might be "special" messages that break this behavior, but that we have not yet seen. This can only be answered by running for a long time and seeing that it does not break.

2) That there is some edge case with how the "last stored chat line" and the  system for how new lines are detected with these ads spanning multiple lines. I cannot say for certain that it will break. But I *can* say that when I wrote the code I made the assumption that we got lines split on line breaks, and now that is not true anymore. So the logic needs to be scrutinized. And I think we need to make some practical tests, where we try to fill the chat so let's say just haft the first welcome or advertising message is left in the chat history, and then we go talk to an NPC, and see what happens.

3) There might be some point 3 as well, but I can't put my finger on it yet...

magicus commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
Let me phrase things differently: I am not overly confident in the stability of this code. In contrast with e.g. the Container Query which I have a lot of faith in, this was tricky to get right, I made a lot of false starts and had to fix things when developing the code that I did not expect when I wrote it. So in my eyes it is very brittle. And I worry that this change will make it even more brittle. When I instead would like to make it more robust.

magicus commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
It is not about optimizing storage, it is about storing the history of the chat on a per-line basis, since minecraft buffers N lines (I think it is 32, or something like that), and that the check-backwards-for-actually-new-messages work on a per-line basis. The disconnect between actual lines as a basic type (which my original code had) and siblings as a basic type, which this PR has, makes me worried that things can go wrong in ways that is hard to predict and hard to test.

magicus commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
To put it bluntly:
```
        List<String> lines = new LinkedList<>(Arrays.asList(msg.split("\\n")));
```
was true, but

```
        List<Component> lines = new LinkedList<>(message.getSiblings());
```

is a lie. This is not a list of lines. It is a list of siblings. And that will have consequences. I like the idea to keep the messages as Components. But I'd *also* like to keep the idea that `lines` are, in fact, lines.

magicus commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
You resolved this without fixing?

magicus commented on common/src/main/java/com/wynntils/features/user/TranslationFeature.java:
Also, to be clear, I think the name "nonBlocking" (or perhaps even better if we switch it around and use "blocking" (which then will be true for the "traditional" NPC dialogues) should be propagated everywhere, into this event and where the event is read, etc.

kristofbolyai commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
Ah, I renamed some other boolean with the wrong terminology.

kristofbolyai commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
I agree that this might not be the most stable/bug-free code we have. However, parsing chat data will always be somewhat uncertain. However, I do think that this PR is not that big of a downgrade to our code in terms of safety/stability. 

Yes, I am a bit paranoid about merging this PR. But, I don't think we can do this much better. Basically, the logic is quite simple now. Is current like a NPC Dialog, according to the pattern, or is it not? The other code is just handling the problem of Wynn re-sending messages.

kristofbolyai commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
Overall, I am not even sure how the component should look like when we "squash" it.

<img width="489" alt="image" src="https://user-images.githubusercontent.com/49001742/191846028-5fd70fc3-bb12-4d6e-a780-32a9f9c3697c.png">

What do we do with the spaces that make this centered? Do we count the first line as line 1, and the 2-3 as a line again? I think you can see what the problem is. We cannot really turn this message into 1 line, without making assumptions. And, one of our goal is to have less assumptions in the code. So we just fix a problem by creating essentially the same problem somewhere else.

magicus commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
I've started trying to write a method for doing this reshuffling of the TextComponent.
Basically what I need to be happy is that this line is replaced with a 
`List<Component> lines = getLinesAsSiblings(message);`

I can help trying to write `getLinesAsSiblings` if you want.

kristofbolyai committed:
Fix up terminology

kristofbolyai commented on common/src/main/java/com/wynntils/features/user/TranslationFeature.java:
Should be renamed everywhere.

kristofbolyai commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
After writing the code, and having proper component - line relationship, do you want to put back the newline checks (is it even necessary if every quest message is its own sibling)?

kristofbolyai commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
Writing the code is half the problem (obviously, it is tricky). I am just not sure this is any better (as I assume it won't be worse. Although, in my head, the more modifications we do to the strings, the more bugs we will introduce).

magicus commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
I found another example of a message that do not follow the "one sibling per line rule":

```
[22:44:03] [Render thread/INFO] (Minecraft) [CHAT] §3[INFO§3]§b TheNelston rewarded §r§31024 Emeralds§r§b to TheNelston.
[22:44:03] [Render thread/INFO] (Minecraft) [CHAT] §r§3Rewards can be claimed in the Member Menu.
```

This was sent as a single sibling.

kristofbolyai commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
You mentioned your concern was that, when writing the ChatModel code, you had the assumption that the `lines` list always contains 1 chat line.

What if, we remove this assumption, and re-factor the ChatModel around this (although, I am not sure if there is anything to rewrite), instead of messing with components?

kristofbolyai commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
> I found another example of a message that do not follow the "one sibling per line rule":
> 
> ```
> [22:44:03] [Render thread/INFO] (Minecraft) [CHAT] §3[INFO§3]§b TheNelston rewarded §r§31024 Emeralds§r§b to TheNelston.
> [22:44:03] [Render thread/INFO] (Minecraft) [CHAT] §r§3Rewards can be claimed in the Member Menu.
> ```
> 
> This was sent as a single sibling.

Seems like relevant messages are "grouped" together. Maybe we could even use this whole sibling - parent relation to our benefit? Example: Let's say we want to filter out the ad. We check for the first line, and we can just remove it's siblings as well. Same with the guild rewards. No need for multiple regexes.

magicus commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
Yes, but what happens the first line scroll of the screen, but the second is still there? And you enter a NPC dialogue...

magicus commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
The chat system of Vanilla is still line-based. That's a hard fact that we need to relate to, and it determines how we must approach this, imho.

kristofbolyai commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
> Yes, but what happens the first line scroll of the screen, but the second is still there? And you enter a NPC dialogue...

I think Wynn either sends the whole component again (and mc just forgets it) or neither is sent. I don't think we confirmed this is not the case?

magicus commented on common/src/main/java/com/wynntils/features/user/overlays/NpcDialogueOverlayFeature.java:
Either I'm confused, or you are confused. Pressing "sneak" is only relevant for blocking NPC dialogues, right? The ones that are not blocking is just announced to the chat.

So the autoProgress scheme in NpcDialogueOverlayFeature is only releavant for `e.isBlocking()`. So I think this test is still backwards.

magicus commented on common/src/main/java/com/wynntils/features/user/overlays/NpcDialogueOverlayFeature.java:
And here as well. `currentlyBlocking` is set to true if we get a blocking NPC dialogue, i.e. one which waits for sneak. So if, and only if, the dialog we are showing is blocking, then we should print this text message.

kristofbolyai commented on common/src/main/java/com/wynntils/features/user/overlays/NpcDialogueOverlayFeature.java:
Blocking is a weird word in this scenario. I think of blocking as "this dialouge blocks the user from progressing by shift". You can think of it as "this dialogue blocks progress until shift"

kristofbolyai commented on common/src/main/java/com/wynntils/features/user/overlays/NpcDialogueOverlayFeature.java:
perhaps this is not that good of a name after all?

magicus commented on common/src/main/java/com/wynntils/features/user/overlays/NpcDialogueOverlayFeature.java:
You do not need to set this here. `currentlyBlocking` is only checked in `renderDialogue`, and that is only called if we have a `currentDialgue`, and that is set at the same time as you set `currentlyBlocking` in the event handler. So this is just confusing. (Especially as you set it to true)

magicus commented:
Another question: How do you know when to clear the non-blocking NPC dialogues from the overlay? I can't really see that this is done now, so I assume they will stay on screen until a new dialogue comes into chat? That might be annoying...

kristofbolyai commented:
> Another question: How do you know when to clear the non-blocking NPC dialogues from the overlay? I can't really see that this is done now, so I assume they will stay on screen until a new dialogue comes into chat? That might be annoying...

Yeah, but most of the time when this happens, there is some animation. I guess I could use the same system for calculating a display time as auto progress?

magicus commented on common/src/main/java/com/wynntils/features/user/overlays/NpcDialogueOverlayFeature.java:
Ok, seems we interpret "blocking" completely different here then! Yes, to  me, it "obviously" meant that progress was blocked, waiting for shift. :)

`needsConfirmation`? Is that ok? So the "normal" NPC dialogues that require you to press shift has `needsConfirmation` true?

magicus commented:
Have we not had this discussion before, in discord or a bug report? I'm thinking of quests where someone is calling out lines, one by one, that explain a part of the mystery, while you traverse further and further into the maze leading to the boss battle. Like https://wynncraft.fandom.com/wiki/The_Hunger_of_Gerts_Part_2_(1.20)#Stage_12

magicus commented:
Trying to parkour or fight mobs with a NPC dialog plastered over your screen might be ... non-helpful....

kristofbolyai committed:
Terminology refactor

kristofbolyai committed:
NpcDialogueOverlayFeature: Auto progress for no confirmation dialogues

kristofbolyai committed:
Merge branch 'main' into clickable_dialogue_fix

kristofbolyai commented:
@magicus did the renames, and added auto progress. 

Are you willing to accept this PR as is? If not, what are the minimum changes you request? 

I think, this is fine for the time being. We can come back to this and figure out a nice solution (if there is one...).

magicus commented:
I am working on a `List<Component> splitComponentInLines(Component message)` method. It's almost done but there are some details left to flesh out.

But I think I can just inject that in your patch once it's pushed. I think that will alleviate my fears. 

Let me double check the patch for any other things, and then I think you're good to go.

magicus commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
Ooh, I did not spot this before. You cannot do this. Not all "NPC" dialogue lines are formatted like that. (That was why I kept the blank lines before, to be able to tell when the "NPC dialogue" lines were done. 

For an example, go and talk to the Detlas citizen standing next to the entrance from Nivla/Ragni. The dialogue ends like "You walk away, wondering if the citizen stands there all day greeting visitors" or something like that.

magicus commented on common/src/main/java/com/wynntils/features/user/TranslationFeature.java:
Should as in "I agree, I need to fix this", or "this should have been fixed"? Because if it's the latter, that does not seem to be the case, or there is something weird with what github shows.

magicus commented on common/src/main/java/com/wynntils/features/user/overlays/NpcDialogueOverlayFeature.java:
I feel like I'm being a real PITA tonight, but this still does not seem right. :-)

Overall I think this latest iteration makes most sense, there are two "kinds" of things we can schedule, an "press sneak for me automatically" and one "remove this overlay after a while". But. The text you add here applies only to the former, since "Press Y to cancel" does not make sense in the latter situation.

Or, at least, this requires some more thought. What would it mean to cancel? Perhaps you need extra time to read it. But if you cancel it, can you "restart" the timer that will clear the message? How? Or is that just making things complicated?

And if you missed the message, how can you re-read it? I don't think these "direct to chat" messages are stored in the quest book history, so if you're a slow reader, it's just lost.

magicus commented:
I created a splitComponentInLines() method in https://github.com/Wynntils/Artemis/pull/429. If you include it, I think you can properly just start replacing String with Component as you do.

However, as I tried to explain, I don't think you can drop all blank lines, since they are needed for proper detection of dialogue lines.

So I think this PR will in the end boil down to:
1) keeping the general structure on how the "chat screens" are segmented into parts
2) ... but use proper Components instead of coded strings
3) ... and introduce an additional check if a chat message is a "free-floating, no-confirmation-needed" dialogue, which 
4) ... is shown by the NPC Overlay in a suitable way.

magicus commented on common/src/main/java/com/wynntils/features/user/overlays/NpcDialogueOverlayFeature.java:
I just had a revelation: Perhaps we can solve all the above problems by this behavior: For a scheduled auto progress for non-need-confirmation messages, the text says "Auto-Progress: .... Press Y to move to chat", and pressing Y will then cancel the NPC Dialogue overlay, but at the same time send out the dialogue to the chat. This way you can either give yourself more time to read it, and/or you can get it out of the way.

kristofbolyai commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
Why? If we do not have a parsable quest, we set current dialogue to be empty. ...or is the problem that we don't parse the "you walk away.." part? It is not an actual dialogue though, is it?

kristofbolyai committed:
Merge branch 'main' into clickable_dialogue_fix

magicus commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
No, we have not tested this. Or at least, I have not.

magicus commented on common/src/main/java/com/wynntils/core/chat/ChatModel.java:
It is treated as part of the dialogue. You need to press shift to get past it. Go do Detlas and check with your code. I don't think it will work.

kristofbolyai commented:
@magicus Do you plan to implement the new line parser or should I do it?

magicus commented:
To be honest, I think this PR should be closed, and that we open two new ones. One in which we switch String to Component, and one in which we catch the no-confirmation-needed messages. It's just gotten so messy by now. I can do the former; after all, all the hard work was in creating the splitComponentInLines. And then you can pick out the "free-floating" NPC dialoges and make a separate PR out of it. Sounds OK?

kristofbolyai commented:
> To be honest, I think this PR should be closed, and that we open two new ones. One in which we switch String to Component, and one in which we catch the no-confirmation-needed messages. It's just gotten so messy by now. I can do the former; after all, all the hard work was in creating the splitComponentInLines. And then you can pick out the "free-floating" NPC dialoges and make a separate PR out of it. Sounds OK?

Yes, good idea.
